### PR TITLE
#7249 - Archived entities should always be read-only

### LINKED ISSUE #7249: Archived entities should always be read-only [3]
### Situation Description
Currently entities are either archived automatically or by an admin. Those archived entites are moved to another view in the directory, but can still be opened and edited by most users of the web application.

### Change Description
- [x] Archived entities should not be editable any more -> open as read-only
- [x] Add a check in the backend persistence methods
- [x] Adjust the German translation of archiving used for core entities (not infrastructure) to 
   * Archive (action) -> "Abschließen"
   * De-archive (action) -> "Wiederereröffnen"
   * archived (status) -> "abgeschlossen"
- [x] Add a feature configuration to disable "archived entities can be not be edited" 

### Additional information
- Remark: We also had "Sperren"("Lock") in mind, but I think this sounds too much like a functionality to temporarily lock the entity for editing by other users. 
- Out of scope: We might add a "prevent sync"-Functionality in the future.
- Defined as `change` on the "archive" feature, this ticket is not a new feature by itself.

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #7249

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/caze/CaseFacade.java [hunk 1]
```diff
@@ -165,7 +165,7 @@ Long countCasesForMap(
 
 	FollowUpPeriodDto calculateFollowUpUntilDate(CaseDataDto caseDto, boolean ignoreOverwrite);
 
-	boolean isCaseEditAllowed(String caseUuid);
+	boolean isCaseEditAllowed(String caseUuid, boolean withArchive);
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/java/de/symeda/sormas/api/caze/CaseFacade.java:168]: @sergiupacurariu Why do we need this parameter? Shoudn't this always be checked?

Comment (MateStrysewske) [reply sormas-api/src/main/java/de/symeda/sormas/api/caze/CaseFacade.java:168]: @MartinWahnschaffe I've already talked to Sergiu about this parameter and proposed an alterantive solution that he'll implement :)

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package de.symeda.sormas.ui.utils;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.vaadin.ui.Component;
+import com.vaadin.ui.CustomLayout;
+
+public class SidePanelLayout extends CustomLayout {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java:10]:
```suggestion
public class LayoutWithSidePanel extends CustomLayout {
```

=> "SidePanelLayout" sounds like it IS a side panel, not like it includes one

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+	public void addSidePanelComponent(Component component, String loc) {
+		sidePanel.addComponent(component, loc);
+	}
+
+	public CustomLayout getSidePanelComponent() {
+		return sidePanel;
+	}
+
+	public void disable(String... excludedButtons) {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java:52]:
```suggestion
	public void disable(String... excludedButtons) {
```
Please also add a comment here to explain what the parameter does

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 4]
```diff
@@ -276,7 +279,7 @@ bulkTaskPriority=Priorität ändern
 # Campaign
 campaignActiveCampaigns=Aktive Kampagnen
 campaignAllCampaigns=Alle Kampagnen
-campaignArchivedCampaigns=Archivierte Kampagnen
+campaignArchivedCampaigns=Abgeschlossen Kampagnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:282]: changed

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 5]
```diff
@@ -346,7 +349,7 @@ caseFacilityDetailsShort=Name der Einrichtung
 caseNewCase=Neuer Fall
 casePlaceOfStay=Aufenthaltsort
 caseActiveCases=Aktive Fälle
-caseArchivedCases=Archivierte Fälle
+caseArchivedCases=Abgeschlossen Fälle
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:352]: changed

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 6]
```diff
@@ -670,7 +673,7 @@ contactResumeFollowUp=Nachverfolgung fortsetzen
 contactSelect=Kontakt auswählen
 contactCreateNew=Neuen Kontakt erstellen
 contactActiveContacts=Aktive Kontakte
-contactArchivedContacts=Archivierte Kontakte
+contactArchivedContacts=Abgeschlossen Kontakte
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:676]: changed

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 7]
```diff
@@ -1069,10 +1072,10 @@ DocumentTemplate.notUploaded=Dokumente konnten nicht in die folgenden Entitäten
 
 # Event
 eventActiveEvents=Aktive Ereignisse
-eventArchivedEvents=Archivierte Ereignisse
+eventArchivedEvents=Abgeschlossen Ereignisse
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:1075]: changed

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 8]
```diff
@@ -1218,7 +1221,7 @@ eventParticipantSelect=Ereignisteilnehmer auswählen
 eventParticipantCreateNew=Neuen Ereignisteilnehmer erstellen
 eventParticipantActiveEventParticipants=Aktive Ereignisteilnehmer
 eventParticipantAllEventParticipants=Alle Ereignisteilnehmer
-eventParticipantArchivedEventParticipants=Archivierte Ereignisteilnehmer
+eventParticipantArchivedEventParticipants=Abgeschlossen Ereignisteilnehmer
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:1224]: changed

File: sormas-api/src/main/resources/captions_de-DE.properties [hunk 9]
```diff
@@ -2321,7 +2324,7 @@ travelEntryOnlyEntriesTestedNegative=Nur negativ getestete Einreisende
 travelEntryOnlyEntriesConvertedToCase=Nur in Fälle konvertierte Einreisen
 travelEntryOpenResultingCase=Fall zu dieser Einreise öffnen
 travelEntryActiveTravelEntries=Aktive Einreisen
-travelEntryArchivedTravelEntries=Archivierte Einreisen
+travelEntryArchivedTravelEntries=Abgeschlossen Einreisen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/captions_de-DE.properties:2327]: changed

File: sormas-api/src/main/resources/strings.properties [hunk 10]
```diff
@@ -356,6 +356,8 @@ errorExternalSurveillanceToolEventNotOwned=%d of the selected events can not be
 errorEventFromAnotherJurisdiction = The event is related to a jurisdiction you don't have access to
 errorEventUnlinkEventGroupFromAnotherJurisdiction = The event group has an event related to a jurisdiction you don't have access to
 errorCaseNotEditable = This case is not editable any more
+errorEntityNotEditable = This entity is not editable any more
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:359]: changed

File: sormas-api/src/main/resources/strings.properties [hunk 11]
```diff
@@ -920,6 +922,7 @@ messageCasesArchived = All selected cases have been archived
 messageCasesDearchived = All selected cases have been de-archived
 messageCasesDeleted = All selected cases have been deleted
 messageCasesEdited = All cases have been edited
+messageCasesBulkEdited = The already archived cases have not been edited. From the total of %s cases submitted for bulk edit %s cases have been edited.
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:925]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 12]
```diff
@@ -898,16 +898,16 @@ messageAllContactsAlreadyInEvent = Alle Kontakte sind bereits mit dem ausgewähl
 messageAllContactsLinkedToEvent = Alle Kontakte wurden mit dem ausgewählten Ereignis verknüpft.
 messageAlreadyEventParticipant = Die Person, die Sie ausgewählt haben, ist bereits als Ereignisteilnehmer dieser Veranstaltung definiert.
 messageAnimalContactsHint = <i>Bitte geben Sie eine Antwort in Bezug auf ALLE Tiere (lebendig oder tot) an, denen die Person während der Inkubationszeit direkt ausgesetzt war (z.B. Jagd, Berühren, Essen).</i>
-messageArchiveUndoneReasonMandatory = Bitte geben Sie einen Grund für die Dearchivierung an
-messageCampaignArchived = Kampagne wurde archiviert
+messageArchiveUndoneReasonMandatory = Bitte geben Sie einen Grund für die Wiedereröffnung an
+messageCampaignArchived = Kampagne wurde abgeschlossen
 messageCampaignCreated = Neue Kampagne erstellt
-messageCampaignDearchived = Kampagne wurde de-archiviert
+messageCampaignDearchived = Kampagne wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:904]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 13]
```diff
@@ -898,16 +898,16 @@ messageAllContactsAlreadyInEvent = Alle Kontakte sind bereits mit dem ausgewähl
...
 messageCampaignDeleted = Kampagne gelöscht
 messageCampaignSaved = Kampagne gespeichert
 messageCampaignFormSaved = %s gespeichert
-messageCaseArchived = Fall wurde archiviert
+messageCaseArchived = Fall wurde abgeschlossen
 messageCaseCreated = Neuer Fall erstellt
-messageCaseDearchived = Fall wurde de-archiviert
+messageCaseDearchived = Fall wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:910]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 14]
```diff
@@ -938,8 +938,8 @@ messageConvertEventParticipantToCase=Sie haben gerade ein positives Laborergebni
 messageConvertEventParticipantToCaseDifferentDiseases=Sie haben gerade ein positives Laborergebnis für eine andere Krankheit gespeichert als die Ereigniskrankheit. Möchten Sie einen Fall mit dieser Krankheit für die Ereignisteilnehmer Person erstellen? Der Fall wird nicht mit dem Ereignisteilnehmer verknüpft.
 messageConvertEventParticipantToCaseNoDisease=Sie haben gerade ein positives Laborergebnis für ein Ereignis ohne Krankheit gespeichert. Möchten Sie einen Fall mit dieser Krankheit für die Ereignisteilnehmer Person erstellen? Das Endergebnis der Probe wird automatisch auf positiv gesetzt, aber der Fall wird nicht mit dem Ereignisteilnehmer verbunden.
 messageContactCreated=Neuer Kontakt angelegt
-messageContactArchived = Kontakt wurde archiviert
-messageContactDearchived = Kontakt wurde dearchiviert
+messageContactArchived = Kontakt wurde abgeschlossen
+messageContactDearchived = Kontakt wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:942]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 15]
```diff
@@ -971,10 +971,10 @@ messageEventSuperordinateEventUnlinked = Die Verbindung zwischen diesem Ereignis
 messageEventSubordinateEventUnlinked = Die Verbindung zwischen diesem Ereignis und seinem untergeordneten Ereignis wurde erfolgreich entfernt
 messageEventParticipationUnlinked = Die Verknüpfung zwischen diesem Fall und dem Ereignis wurde erfolgreich getrennt
 messageEventUnlinkedFromEventGroup = Die Verknüpfung zwischen diesem Ereignis und der Ereignisgruppe wurde erfolgreich getrennt
-messageEventDearchived = Ereignis wurde de-archiviert
+messageEventDearchived = Ereignis wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:974]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 16]
```diff
@@ -971,10 +971,10 @@ messageEventSuperordinateEventUnlinked = Die Verbindung zwischen diesem Ereignis
...
 messageEventGroupDearchived = Ereignisgruppe wurde de-archiviert
-messageEventParticipantArchived = Ereignisteilnehmer wurde archiviert
-messageEventParticipantDearchived = Ereignisteilnehmer wurde dearchiviert
+messageEventParticipantArchived = Ereignisteilnehmer wurde abgeschlossen
+messageEventParticipantDearchived = Ereignisteilnehmer wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:977]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 17]
```diff
@@ -1004,8 +1004,8 @@ messageFacilityChanged = Sie haben die Einrichtung dieses Falls geändert. Möch
 messageFacilityMulitChanged = Sie haben die Einrichtung mehrerer Fälle geändert. Möchten Sie den Fall an die neue Einrichtung übertragen (der Krankenhausaufenthalt wird möglicherweise aktualisiert) oder nur die Daten bearbeiten, um einen Fehler zu korrigieren?
 messageGdpr = Gemäß dem von der DSGVO vorgeschriebenen Prinzip der Datenminimierung müssen Sie sicherstellen, dass Sie hier nur objektive Daten eingeben, die für die Verarbeitung der Daten absolut notwendig sind. Insbesondere dürfen Sie weder biometrische, philosophische, politische oder religiöse Daten, noch gewerkschaftliche Meinungen, sexuelles Leben oder Orientierung, ethnische Herkunft, Straftaten, Verurteilungen, Sicherheitsmaßnahmen, Vorbestrafungen ... oder andere Daten, die nicht mit der Gesundheitsbefragung zusammenhängen eingeben.
 messageGdprCheck = Ich habe diese Informationen gelesen, bitte dieses Fenster nicht mehr anzeigen
-messageImmunizationArchived = Immunisierung wurde archiviert
-messageImmunizationDearchived = Immunisierung wurde dearchiviert
+messageImmunizationArchived = Immunisierung wurde abgeschlossen
+messageImmunizationDearchived = Immunisierung wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:1008]: changed

File: sormas-api/src/main/resources/strings_de-DE.properties [hunk 18]
```diff
@@ -1067,8 +1067,8 @@ messageTasksDearchived = Alle ausgewählten Aufgaben wurden de-archiviert
 messageTasksDeleted = Alle ausgewählten Aufgaben wurden gelöscht
 messageTemplateNotAvailable = Die Vorlagen-Datei ist nicht verfügbar. Bitte kontaktieren Sie einen Administrator und teilen Sie ihm dieses Problem mit.
 messageTravelEntrySaved = Einreisedaten gespeichert
-messageTravelEntryArchived = Einreise wurde archiviert
-messageTravelEntryDearchived = Einreise wurde dearchiviert
+messageTravelEntryArchived = Einreise wurde abgeschlossen
+messageTravelEntryDearchived = Einreise wurde wiedereröffnen
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings_de-DE.properties:1071]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CasesView.java [hunk 19]
```diff
@@ -690,12 +690,12 @@ public HorizontalLayout createStatusFilterBar() {
 									items -> ControllerProvider.getCaseController().sendSmsToAllSelectedItems(items, () -> navigateTo(criteria)));
 							}, hasBulkOperationsRight));
 					}
-					menuBarItems.add(new MenuBarHelper.MenuBarItem(I18nProperties.getCaption(Captions.actionArchive), VaadinIcons.ARCHIVE, mi -> {
+					menuBarItems.add(new MenuBarHelper.MenuBarItem(I18nProperties.getCaption(Captions.actionArchiveInfrastructure), VaadinIcons.ARCHIVE, mi -> {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CasesView.java:693]: changed to ActionArchiveCoreEntity

File: sormas-ui/src/main/java/de/symeda/sormas/ui/events/EventsView.java [hunk 20]
```diff
@@ -552,15 +552,15 @@ public HorizontalLayout createStatusFilterBar() {
 				if (UserProvider.getCurrent().hasUserRight(UserRight.EVENT_ARCHIVE)) {
 					bulkActions.add(
 						new MenuBarHelper.MenuBarItem(
-							I18nProperties.getCaption(Captions.actionArchive),
+							I18nProperties.getCaption(Captions.actionArchiveInfrastructure),
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/events/EventsView.java:555]: changed to ActionArchiveCoreEntity

File: sormas-ui/src/main/java/de/symeda/sormas/ui/task/TaskGridComponent.java [hunk 21]
```diff
@@ -191,14 +191,14 @@ public HorizontalLayout createAssigneeFilterBar() {
 						hasBulkOperationsRight));
 				menuBarItems.add(
 					new MenuBarHelper.MenuBarItem(
-						I18nProperties.getCaption(Captions.actionArchive),
+						I18nProperties.getCaption(Captions.actionArchiveInfrastructure),
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/task/TaskGridComponent.java:194]: changed to ActionArchiveCoreEntity

[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/task/TaskGridComponent.java:194]: for all above ActionDearchiveInfrastructure has been replaced with actionDearchiveCoreEntity

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/CommitDiscardWrapperComponent.java [hunk 22]
```diff
@@ -826,4 +828,15 @@ public boolean isDirty() {
 	public void setDirty(boolean dirty) {
 		this.dirty = dirty;
 	}
+
+	public void setEditable(boolean editable, String... buttonsExcepted) {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/CommitDiscardWrapperComponent.java:832]: added

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java:10]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java (hunk 3 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/SidePanelLayout.java:52]: explanation added

**Comment by MateStrysewske** [comment]
@sergiupacurariu Code looks good to me, please find someone to test the changes :)

**Review by alexcaruntu-vita — APPROVED** [review]

**Pull request merged**